File: python-backend/mineforgeai/agent/shell.py
from __future__ import annotations

import re


DANGEROUS_PATTERNS = [
    re.compile(r"rm\s+-rf\s+\*", re.IGNORECASE),
    re.compile(r"rm\s+-rf\s+/$", re.IGNORECASE),
    re.compile(r"del\s+/s", re.IGNORECASE),
    re.compile(r"format(\s|$)", re.IGNORECASE),
    re.compile(r"diskpart", re.IGNORECASE),
    re.compile(r"shutdown", re.IGNORECASE),
    re.compile(r"reboot", re.IGNORECASE),
    re.compile(r"mkfs", re.IGNORECASE),
    re.compile(r"dd\s+if=", re.IGNORECASE),
    re.compile(r"curl.+\|\s*sh", re.IGNORECASE),
    re.compile(r"wget.+\|\s*sh", re.IGNORECASE),
]


def is_safe_command(command: str) -> bool:
    return not any(pattern.search(command) for pattern in DANGEROUS_PATTERNS)

File: lib-node/shell.js
const DANGEROUS_PATTERNS = [
  /rm\s+-rf\s+\/?$/i,
  /rm\s+-rf\s+\*/i,
  /del\s+\/s/i,
  /format(\s|$)/i,
  /diskpart/i,
  /shutdown/i,
  /reboot/i,
  /mkfs/i,
  /dd\s+if=/i,
  /curl.+\|\s*sh/i,
  /wget.+\|\s*sh/i
];

export function isDangerousCommand(command) {
  return DANGEROUS_PATTERNS.some((pattern) => pattern.test(command));
}

export function getShellCommand(command, platform = process.platform) {
  if (platform === "win32") {
    return { executable: "cmd.exe", args: ["/d", "/s", "/c", command] };
  }
  return { executable: "/bin/sh", args: ["-lc", command] };
}
